File: config/src/main/scala/net/kemitix/thorp/config/SourceConfigLoader.scala
package net.kemitix.thorp.config

import java.io.File
import java.util

import net.kemitix.thorp.domain.Sources

import scala.jdk.CollectionConverters._

trait SourceConfigLoader {

  val thorpConfigFileName = ".thorp.conf"

  def loadSourceConfigs(sources: Sources): ConfigOptions =
    sources
      .paths()
      .asScala
      .toList
      .map(path =>
        ParseConfigFile.parseFile(new File(path.toFile, thorpConfigFileName)))
      .foldLeft(ConfigOptions(sourceConfigOptions(sources)))((acc, co) =>
        acc ++ co)

  private def sourceConfigOptions(
      sources: Sources): java.util.List[ConfigOption] = {
    val options: util.ArrayList[ConfigOption] = new util.ArrayList[ConfigOption]
    sources
      .paths()
      .stream()
      .map(path => ConfigOption.Source(path))
      .forEach(o => options.add(o))
    options
  }
}

object SourceConfigLoader extends SourceConfigLoader

File: config/src/main/java/net/kemitix/thorp/config/SourceConfigLoader.java
package net.kemitix.thorp.config;

import net.kemitix.thorp.domain.Sources;

import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.util.stream.Collectors;

public interface SourceConfigLoader {

    static ConfigOptions loadSourceConfigs(Sources sources) throws IOException {
        // add each source as an option
        ConfigOptions configOptions =
                ConfigOptions.create(
                        sources.paths()
                                .stream()
                                .peek(path -> {
                                    System.out.println("Using source: " + path);
                                })
                                .map(ConfigOption::source)
                                .collect(Collectors.toList()));
        // add settings from each source as options
        for (Path path : sources.paths()) {
            configOptions = configOptions.merge(
                    ParseConfigFile.parseFile(
                            new File(path.toFile(), ".thorp.conf")));
        }
        return configOptions;
    }

}
